nni/nas/oneshot/pytorch/supermodule/proxyless.py

```diff
@@ -68,116 +90,172 @@ def element_product_sum(tensor1: tuple[torch.Tensor, ...], tensor2: tuple[torch.
+        #                   Need to restore the original object structure.
+        # NOTE: This could be potentially troublesome if users actually returned a single-ele tuple.
+        return _unpack_tuple(_detach_tensor(layer_output, requires_grad=True))
+
+    @staticmethod
+    def backward(ctx, *grad_output):
+        softmax, sample_index, forward_path_func, num_inputs = \
+            ctx.softmax, ctx.sample_index, ctx.forward_path_func, ctx.num_inputs
+
+        if ctx.needs_input_grad[0]:
+            # arch_alpha requires gradient. To improve efficiency.
+            grads = None
+        else:
+            arch_alpha, *layer_input_output = ctx.saved_tensors
+            layer_input = layer_input_output[:num_inputs]
+            layer_output = layer_input_output[num_inputs:]
+
+            # binary_grads is the gradient of binary gates.
+            # Binary gates is a one-hot tensor where 1 is on the sampled index, and others are 0.
+            # By chain rule, it's gradient is grad_output times the layer_output (of the corresponding path).
+
+            binary_grads = torch.zeros_like(arch_alpha)
+
+            with torch.no_grad():
+                # Compute binary grads.
+                for k in range(len(binary_grads)):
+                    if k != sample_index:
+                        out_k = forward_path_func(k, *layer_input)  # type: ignore
+                    else:
+                        out_k = layer_output
+
+                    # NOTE: One limitation here is that out_k can't be complex objects like dict.
+                    binary_grads[k] = element_product_sum(
+                        _pack_as_tuple(out_k),  # In case out_k is a single tensor
+                        grad_output             # Assuming grad_output is not None
+                    )
+
+                # Compute the gradient of the arch_alpha, based on binary_grads.
+                grads = torch.zeros_like(arch_alpha)
+                probs = softmax(arch_alpha)
+                for i in range(len(arch_alpha)):
+                    for j in range(len(arch_alpha)):
+                        # Arch alpha's gradients are accumulated for all backwards through this layer.
+                        grads[i] += binary_grads[j] * probs[j] * (int(i == j) - probs[i])
+
+        return (
+            None, None, None, None,     # No gradients for the constants
+            grads,                      # arch_alpha
+            *([None] * num_inputs),     # We only use the inputs. We know nothing about the inputs.
+            *grad_output                # Pass-through
+        )
+
+
+def suppress_already_mutated(module, name, memo, mutate_kwargs) -> bool:
+    # ProxylessMixedRepeat will create MixedLayer on its own.
+    # The created layer should NOT be mutated again.
+
+    if isinstance(module, (ProxylessMixedLayer, ProxylessMixedInput, ProxylessMixedRepeat)):
+        return True
+
+    return None
```

Comment:
bool?

Reply:
Return none to skip the hook. Changed the signature.